# Wrangle Exercises
---
## Exercises I
Let's review the steps we take at the beginning of each new module.

1. Create a new repository named `regression-exercises` in your GitHub; all of your Regression work will be housed here.
2. Clone this repository within your local `codeup-data-science` directory.
3. Create a `.gitignore` and make sure your list of 'files to ignore' includes your `env.py` file.
4. Ceate a `README.md` file that outlines the contents and purpose of your repository.
5. Add, commit, and push these two files.
6. Now you can add your `env.py` file to this repository to access the Codeup database server.
7. For these exercises, you will create `wrangle.ipynb` and `wrangle.py` files to hold necessary functions.
8. As always, add, commit, and push your work often.
---
## Exercises II
Let's set up an example scenario as perspective for our regression exercises using the Zillow dataset.

As a Codeup data science graduate, you want to show off your skills to the Zillow data science team in hopes of getting an interview for a position you saw pop up on LinkedIn. You thought it might look impressive to build an end-to-end project in which you use some of their Kaggle data to predict property values using some of their available features; who knows, you might even do some feature engineering to blow them away. Your goal is to predict the values of single unit properties using the obervations from 2017.

In these exercises, you will complete the first step toward the above goal: acquire and prepare the necessary Zillow data from the zillow database in the Codeup database server.

1. Acquire `bedroomcnt`, `bathroomcnt`, `calculatedfinishedsquarefeet`, `taxvaluedollarcnt`, `yearbuilt`, `taxamount`, and `fips` from the `zillow` database for all 'Single Family Residential' properties.

In [1]:
# import pandas
import pandas as pd
# function to write url for sql database
def get_url(db):
    from env import user, password, host
    return f'mysql+pymysql://{user}:{password}@{host}/{db}'

In [2]:
# get url for zillow database
url = get_url('zillow')
# sql query to acquire data
sql = '''
SELECT bedroomcnt, 
       bathroomcnt,
       calculatedfinishedsquarefeet,
       taxvaluedollarcnt,
       yearbuilt,
       taxamount,
       fips
FROM properties_2017
WHERE propertylandusetypeid = 261;
'''
# assign sql query result (dataframe) to variable
zillow = pd.read_sql(sql, url)
zillow.head()

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  \
0         0.0          0.0                           NaN            27516.0   
1         0.0          0.0                           NaN               10.0   
2         0.0          0.0                           NaN               10.0   
3         0.0          0.0                           NaN             2108.0   
4         4.0          2.0                        3633.0           296425.0   

   yearbuilt  taxamount    fips  
0        NaN        NaN  6037.0  
1        NaN        NaN  6037.0  
2        NaN        NaN  6037.0  
3        NaN     174.21  6037.0  
4     2005.0    6941.39  6037.0

In [3]:
# make function to acquire/cache data
def acquire_zillow():
    '''
    This function takes no arguments and returns a dataframe of 2017 Single Family 
    Residential property data from Zillow. It searches for a csv file (zillow.csv)
    with the requested data and reads that file into a dataframe. If the csv
    file is not found, it retrieves the SQL query result and reads it into a 
    dataframe. It then caches this data into a csv file (zillow.csv).
    '''
    import os
    if os.path.isfile('zillow.csv'):
        zillow = pd.read_csv('zillow.csv', index_col=0)
        return zillow
    else:
        from env import user, password, host
        url = f'mysql+pymysql://{user}:{password}@{host}/zillow'
        sql = '''
        SELECT bedroomcnt, 
               bathroomcnt,
               calculatedfinishedsquarefeet,
               taxvaluedollarcnt,
               yearbuilt,
               taxamount,
               fips
        FROM properties_2017
        WHERE propertylandusetypeid = 261;
        '''
        zillow = pd.read_sql(sql, url)
        zillow.to_csv('zillow.csv')
        return zillow
# check that function works
zillow = acquire_zillow()
zillow.head()

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  \
0         0.0          0.0                           NaN            27516.0   
1         0.0          0.0                           NaN               10.0   
2         0.0          0.0                           NaN               10.0   
3         0.0          0.0                           NaN             2108.0   
4         4.0          2.0                        3633.0           296425.0   

   yearbuilt  taxamount    fips  
0        NaN        NaN  6037.0  
1        NaN        NaN  6037.0  
2        NaN        NaN  6037.0  
3        NaN     174.21  6037.0  
4     2005.0    6941.39  6037.0

2. Using your acquired Zillow data, walk through the summarization and cleaning steps in your `wrangle.ipynb` file like we did above. You may handle the missing values however you feel is appropriate and meaningful; remember to document your process and decisions using markdown and code commenting where helpful.

In [4]:
zillow.shape

(2152863, 7)

In [7]:
zillow.describe().T

count           mean            std  \
bedroomcnt                    2152852.0       3.287196       0.954754   
bathroomcnt                   2152852.0       2.230688       0.999280   
calculatedfinishedsquarefeet  2144379.0    1862.855178    1222.125124   
taxvaluedollarcnt             2152370.0  461896.237963  699676.049600   
yearbuilt                     2143526.0    1960.949681      22.162196   
taxamount                     2148421.0    5634.865978    8178.910249   
fips                          2152863.0    6048.377335      20.433292   

                                  min        25%        50%        75%  \
bedroomcnt                       0.00       3.00       3.00       4.00   
bathroomcnt                      0.00       2.00       2.00       3.00   
calculatedfinishedsquarefeet     1.00    1257.00    1623.00    2208.00   
taxvaluedollarcnt                1.00  188170.25  327671.00  534527.00   
yearbuilt                     1801.00    1949.00    1958.00    1976.00   
taxamount                        1.85    2534.98    4108.95    6414.32   
fips                          6037.00    6037.00    6037.00    6059.00   

                                      max  
bedroomcnt                          25.00  
bathroomcnt                         32.00  
calculatedfinishedsquarefeet    952576.00  
taxvaluedollarcnt             98428909.00  
yearbuilt                         2016.00  
taxamount                      1337755.86  
fips                              6111.00

In [8]:
zillow.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2152863 entries, 0 to 2152862
Data columns (total 7 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   bedroomcnt                    float64
 1   bathroomcnt                   float64
 2   calculatedfinishedsquarefeet  float64
 3   taxvaluedollarcnt             float64
 4   yearbuilt                     float64
 5   taxamount                     float64
 6   fips                          float64
dtypes: float64(7)
memory usage: 115.0 MB


In [34]:
zillow.isna().sum()

bedroomcnt                        11
bathroomcnt                       11
calculatedfinishedsquarefeet    8484
taxvaluedollarcnt                493
yearbuilt                       9337
taxamount                       4442
fips                               0
dtype: int64

In [17]:
zillow[zillow.bedroomcnt.isna()]

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
107763          NaN          NaN                           NaN   
118612          NaN          NaN                           NaN   
193993          NaN          NaN                        1348.0   
1141339         NaN          NaN                         200.0   
1324608         NaN          NaN                         990.0   
1442975         NaN          NaN                           NaN   
1647346         NaN          NaN                         400.0   
1701026         NaN          NaN                           NaN   
1722707         NaN          NaN                           NaN   
1776422         NaN          NaN                           NaN   
1797985         NaN          NaN                           NaN   

         taxvaluedollarcnt  yearbuilt  taxamount    fips  
107763             67366.0     1926.0     780.54  6059.0  
118612             43992.0     1946.0     541.64  6059.0  
193993            840698.0     1952.0        NaN  6059.0  
1141339           188972.0        NaN        NaN  6037.0  
1324608           435000.0     1906.0        NaN  6037.0  
1442975           273196.0        NaN        NaN  6037.0  
1647346            28347.0     1954.0        NaN  6037.0  
1701026           407930.0     1926.0        NaN  6037.0  
1722707           477161.0        NaN        NaN  6037.0  
1776422            38855.0        NaN        NaN  6037.0  
1797985          1017957.0     1990.0        NaN  6037.0

Both bedroomcnt and bathroomcnt are missing 11 values; the data above shows me that they are missing values in the same rows. Considering there are over 2 million observations in this dataset, I should be able to drop these 11 rows without affecting the data too much.

This spurs another question: if I were to drop all rows with null values, how much of the data would I actually lose?

I know it would be at least 9337 rows since yearbuilt is missing 9337 values. It is definitely more than that though because I can see that some of the 11 rows depicted above are not missing a yearbuilt value. Next I will try to determine the total number of rows containing null values to see if it would be too significant to drop all of those rows.

In [42]:
# create series of bools for whether or not each row contains a null
zillow.isna().any(axis=1)
# count true values in series
zillow.isna().any(axis=1).sum()

12628

There are 12,628 rows containing null values!

In [45]:
# find percentage of zillow's rows with null values
print(f'Proportion of rows with null values in zillow dataset: {((zillow.isna().any(axis=1).sum())/len(zillow)):.2%}')

Proportion of rows with null values in zillow dataset: 0.59%


Only about half a percent of the zillow dataset would be lost if I were to drop all rows with null values. I think that is reasonable and won't compromise the data, so I will go ahead and drop all rows containing null values.

After dropping all null values from the dataset, I will be left with 99.4% of the original data retrieved from the SQL server.

In [47]:
# drop rows with null values, reassign dataframe variable
zillow = zillow.dropna()
# verify rows were dropped
zillow.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2140235 entries, 4 to 2152862
Data columns (total 7 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   bedroomcnt                    float64
 1   bathroomcnt                   float64
 2   calculatedfinishedsquarefeet  float64
 3   taxvaluedollarcnt             float64
 4   yearbuilt                     float64
 5   taxamount                     float64
 6   fips                          float64
dtypes: float64(7)
memory usage: 130.6 MB


Next, I will convert column data types to int where appropriate.

In [53]:
# list columns to be converted
int_cols = ['bedroomcnt', 'bathroomcnt', 'calculatedfinishedsquarefeet', 'taxvaluedollarcnt', 'yearbuilt', 'fips']
# create for loop to convert each col in list
for col in int_cols:
    zillow[col] = zillow[col].astype(int)
# confirm dtypes have been changed
zillow.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2140235 entries, 4 to 2152862
Data columns (total 7 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   bedroomcnt                    int64  
 1   bathroomcnt                   int64  
 2   calculatedfinishedsquarefeet  int64  
 3   taxvaluedollarcnt             int64  
 4   yearbuilt                     int64  
 5   taxamount                     float64
 6   fips                          int64  
dtypes: float64(1), int64(6)
memory usage: 130.6 MB


3. Store all of the necessary functions to automate your process from acquiring the data to returning a cleaned dataframe with no missing values in your `wrangle.py` file. Name your final function `wrangle_zillow`.

In [5]:
def wrangle_zillow():
    '''
    This function takes in no arguments and returns a clean dataframe of Single Family
    Residential property data from Zillow.
    '''